api/proto/teleport/legacy/types/types.proto

```diff
@@ -873,6 +873,23 @@ message AppSpecV3 {
   repeated string UserGroups = 8;
 }
 
+// AppServerOrSAMLIdPServiceProviderV1 holds either an AppServerV3 or a SAMLIdPServiceProviderV1 resource (never both).
+// This is the resource used for WebUI requests to list Applications since we want to list both AppServers and
+// SAMLIdPServiceProviders in the UI.
+message AppServerOrSAMLIdPServiceProviderV1 {
+  option (gogoproto.goproto_stringer) = false;
+  option (gogoproto.stringer) = false;
+
+  // Kind is the resource kind. Always "app_server_saml_idp_sp".
+  string Kind = 1 [(gogoproto.jsontag) = "kind"];
+  // AppServer is the AppServer resource. If it is nil, it means this AppServerOrSAMLIdPServiceProvider
+  // resource holds a SAMLIdPServiceProvider.
+  AppServerV3 AppServer = 2 [(gogoproto.jsontag) = "app_server,omitempty"];
```

Comment:
Consider using a `oneof` for this?

https://protobuf.dev/programming-guides/proto3/#oneof